cmd/shutdown.go

```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+const (
+	desiredTimeout  = 5 * time.Second
+	shutdownTimeout = 10 * time.Second
+)
+
+//GracefulShutdown implements releasing all resouces it got from system, finish all request handling and return responses when service stopping.
+func gracefulShutdown(done chan<- bool) error {
+	sigs := make(chan os.Signal)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	sig := <-sigs
+	log.Println("Recieved", sig, "signal")
+
+	ctx, cancel := context.WithTimeout(context.Background(), desiredTimeout)
+
+	select {
+	case <-ctx.Done():
+		log.Println(ctx.Err())
+	case <-time.After(shutdownTimeout):
+		log.Println("Overslept")
+		cancel()
+	default:
+		cancel()
+	}
```

Comment:
Can you explain logic of this?

Reply:
Think, I should use that for safety canceling when context expires. Case with Done() is selected when the context passed as an argument (in our occasion - "cancel") notifies completion, I also use use After() to insure we cancel if it takes too much time. But now I see, I don't need default.